packages/remix-dev/compiler/assets.ts

```diff
@@ -112,47 +113,35 @@ export async function createAssetsManifest(
   return { version, entry, routes };
 }
 
-type ImportsCache = { [routeId: string]: string[] };
-
 function optimizeRoutes(
   routes: AssetsManifest["routes"],
   entryImports: string[]
 ): void {
-  // This cache is an optimization that allows us to avoid pruning the same
+  // Memoize is an optimization that allows us to avoid pruning the same
   // route's imports more than once.
-  let importsCache: ImportsCache = Object.create(null);
-
-  for (let key in routes) {
-    optimizeRouteImports(key, routes, entryImports, importsCache);
-  }
-}
-
-function optimizeRouteImports(
-  routeId: string,
-  routes: AssetsManifest["routes"],
-  parentImports: string[],
-  importsCache: ImportsCache
-): string[] {
-  if (importsCache[routeId]) return importsCache[routeId];
-
-  let route = routes[routeId];
+  let optimizeRouteImports = memoize(function (
+    routeId: string,
+    routes: AssetsManifest["routes"],
+    parentImports: string[]
+  ): string[] {
+    let route = routes[routeId];
 
-  if (route.parentId) {
-    parentImports = parentImports.concat(
-      optimizeRouteImports(route.parentId, routes, parentImports, importsCache)
-    );
-  }
+    if (route.parentId) {
+      let optimized = optimizeRouteImports(route.parentId, routes, parentImports);
+      parentImports = parentImports.concat(optimized);
```

Comment:
Let's keep this as it was
```suggestion
      parentImports = parentImports.concat(
        optimizeRouteImports(route.parentId, routes, parentImports)
      );
```